porcupine/plugins/git_status.py

```diff
@@ -0,0 +1,250 @@
+"""Color items in the directory tree based on their git status."""
+from __future__ import annotations
+
+import os
+import logging
+import subprocess
+import sys
+import time
+from functools import partial
+from pathlib import Path
+from concurrent.futures import ThreadPoolExecutor, Future
+from typing import Any, Iterator
+
+from porcupine import get_paned_window, utils
+from porcupine.plugins.directory_tree import DirectoryTree, get_path
+
+setup_after = ["directory_tree"]
+
+log = logging.getLogger(__name__)
+
+# Each git subprocess uses one cpu core
+_git_pool = ThreadPoolExecutor(max_workers=os.cpu_count())
+
+
+def run_git_status(project_root: Path) -> dict[Path, str]:
+    try:
+        start = time.perf_counter()
+        run_result = subprocess.run(
+            ["git", "status", "--ignored", "--porcelain"],
+            cwd=project_root,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,  # for logging error message
+            encoding=sys.getfilesystemencoding(),
+            **utils.subprocess_kwargs,
+        )
+        log.debug(
+            f"running git status in {project_root} took"
+            f" {round((time.perf_counter() - start)*1000)}ms"
+        )
+
+        if run_result.returncode != 0:
+            # likely not a git repo because missing ".git" dir
+            log.debug(f"git status failed in {project_root}: {run_result}")
+            return {}
+
+    except (OSError, UnicodeError):
+        log.warning("can't run git", exc_info=True)
+        return {}
+
+    # Show .git as ignored, even though it actually isn't
+    result = {project_root / ".git": "git_ignored"}
+    for line in run_result.stdout.splitlines():
+        path = project_root / line[3:]
+        if line[1] == "M":
+            result[path] = "git_modified"
+        elif line[1] == " ":
+            result[path] = "git_added"
+        elif line[:2] == "??":
+            result[path] = "git_untracked"
+        elif line[:2] == "!!":
+            result[path] = "git_ignored"
+        elif line[:2] in {"AA", "UU"}:
+            result[path] = "git_mergeconflict"
+        else:
+            log.warning(f"unknown git status line: {repr(line)}")
+    return result
+
+
+# TODO: copy/pasted from directory_tree plugin
+def _path_to_root_inclusive(path: Path, root: Path) -> Iterator[Path]:
```

Comment:
Used in only one function, consider rewriting that

In directory_tree.py it is also used in exactly one place and that might benefit from some rewriting too